content/unroll-default-arguments.md

```diff
@@ -0,0 +1,817 @@
+make a huge difference in most cases, but it could be useful in scenarios where there
+are a large number of default parameters being added every version, as it would minimize the
+amount of generated code.
+
+
+### Abstract Methods
+
+In [Limitations](#limitations), I mentioned that `@unroll` only supports `final` methods. 
+It is likely possible for abstract methods which are `@unrolled` to have concrete forwarder
+methods generated on their behalf.
+
+```scala
+import scala.annotation.unroll
+
+trait Unrolled{
+  def foo(s: String, n: Int = 1, @unroll b: Boolean = true): String
+}
+
+object Unrolled extends Unrolled{
+  def foo(s: String, n: Int = 1, b: Boolean = true) = s + n + b
+}
+```
+
+Unrolls to:
+
+```scala
+trait Unrolled{
+  def foo(s: String, n: Int = 1, @unroll b: Boolean = true): String = foo(s, n)
+  def foo(s: String, n: Int = 1): String = foo(s, n, true)
+}
+
+object Unrolled extends Unrolled{
+  def foo(s: String, n: Int = 1, b: Boolean = true) = s + n + b
+}
+```
+
+As the forwarders are concrete, the implementor of the abstract method does
+not need to `@unroll` the implementation: they only need to provide the implementation
+for the primary method `def` and not the forwarders.
+
+One thing to note is that the `@unroll`ed abstract method needs to _itself_ become a
+forwarder method, despite originally being abstract! That is because downstream code
+compiled against an old version may define classes which `extends Unrolled` and
+define a concrete `def foo(s: String, n: Int = 1): String`, while other downstream code compiled
+against a newer version may define a concrete
+`def foo(s: String, n: Int = 1, b: Boolean = true): String`. Thus, we need both overloads
+of `foo` to be forwarders, so that downstream code can override either version and still work.
+
+This handling for abstract methods is not fully fleshed out or implemented, so I'm
+not sure if it can truly be made to work.
+
+### Should the Generated Methods be Deprecated or Invisible?
+
+It is not clear to me if we should discourage usage of the generated forwarders 
+methods, or hide them:
+
+1. On one hand, downstream code should not be compiling against these methods: they are 
+   purely for binary compatibility
+
+2. On the other hand, downstream code does not control which overload of the method is 
+   selected: that is up to the Scala compiler and how overloading interacts with default
+   parameter values. I have encountered scenarios with manually-written forwarders where
+   selected over the "primary" implementation
+
+3. The forwarders are meant to have the exact same semantics as the "primary" implementation.
+   Thus it _does not matter_ whether you call the primary and pass it a default value,
+   or you call a forwarder which then calls the primary and passes it the same default value.
+
+For now, I have left the generated methods "as is", though choosing to hide them or deprecate
+them or something else is definitely an option
```

Comment:
Visible synthetic overloads would affect the user experience in IDEs, scaladocs, etc.

Reply:
👍

Are macro annotation generated methods also hidden in separate compilation?

Reply:
They are.